server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java

```diff
@@ -68,147 +66,171 @@
-                .filter(e -> invalidMessagesTester.test(e.getValue()))
-                .map(e -> new AbstractMap.SimpleEntry<>(e.getKey(), buildSetErrorFromValidationResult(e.getValue().validate())))
-                .collect(Collectors.toMap(AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
+    private Either<ErrorWithId, MessageWithId> processEntry(CreationMessageEntry entry, MailboxSession mailboxSession) {
+        return Either.<ErrorWithId, CreationMessageEntry>right(entry)
+            .joinRight(x -> notImplementedErrors(x, mailboxSession))
+            .joinRight(x -> getInvalidArgumentsErrors(x))
+            .joinRight(x -> handleOutboxMessages(x, mailboxSession));
     }
 
-    private SetError buildSetErrorFromValidationResult(List<ValidationResult> validationErrors) {
-        String formattedValidationErrorMessage = validationErrors.stream()
-                .map(err -> err.getProperty() + ": " + err.getErrorMessage())
-                .collect(Collectors.joining("\\n"));
-        Splitter propertiesSplitter = Splitter.on(',').trimResults().omitEmptyStrings();
-        Set<MessageProperties.MessageProperty> properties = validationErrors.stream()
-                .flatMap(err -> propertiesSplitter.splitToList(err.getProperty()).stream())
-                .flatMap(MessageProperty::find)
-                .collect(Collectors.toSet());
-        return SetError.builder()
-                .type("invalidProperties")
-                .properties(properties)
-                .description(formattedValidationErrorMessage)
-                .build();
+    private Either<ErrorWithId, CreationMessageEntry> notImplementedErrors(CreationMessageEntry entry, MailboxSession session) {
+        if (!isMessageSetIn(Role.DRAFTS, entry.getValue(), session)) {
+            return Either.right(entry);
+        } else {
+            return Either.left(
+                    new ErrorWithId(
+                            entry.getCreationId(), 
+                            SetError.builder().type("error").description("Not yet implemented").build()));
+        }
+    }
+    
+    private Either<ErrorWithId, CreationMessageEntry> getInvalidArgumentsErrors(CreationMessageEntry entry) {
+        CreationMessage message = entry.getValue();
+        if (!message.isValid()) {
+            return Either.left(new ErrorWithId(entry.getCreationId(), buildSetErrorFromValidationResult(message.validate())));
+        }
+        return Either.right(entry);
     }
+    
+    private Either<ErrorWithId, MessageWithId> handleOutboxMessages(CreationMessageEntry entry, MailboxSession session) {
+        CreationMessageId creationId = entry.getCreationId();
 
+        return getMailboxWithRole(session, Role.OUTBOX)
+            .map(outbox -> {
+                if (isRequestForSending(entry.getValue(), session)) {
+                    Function<Long, MessageId> idGenerator = uid -> generateMessageId(session, outbox, uid);
```

Comment:
Do you really need to have this function as a parameter? You have other implementations?